src/hotspot/share/opto/vectorization.hpp

```diff
@@ -667,289 +698,253 @@ class VLoopAnalyzer : StackObj {
-  //                    know both address regions inside the same memory object, but these regions are non-
-  //                    overlapping), or the VPointers point to entirely different objects.
-  //                    -> No memory edge required. Aliasing impossible.
-  //
-  //  - Equal:          Synonymous to "overlap, or point to different memory objects".
-  //                    The two VPointers either overlap on the same memory object, or point to two different
-  //                    memory objects.
-  //                    -> Memory edge required. Aliasing likely.
-  //
-  // In a future refactoring, we can simplify to two states:
-  //  - NeverAlias:     instead of Less / Greater
-  //  - MayAlias:       instead of Equal / NotComparable
-  //
-  // Two VPointer are "comparable" (Less / Greater / Equal), iff all of these conditions apply:
-  //   1) Both are valid, i.e. expressible in the compound-long-int or simple form.
-  //   2) The adr are identical, or both are array bases of different arrays.
-  //   3) They have identical scale.
-  //   4) They have identical invar.
-  //   5) The difference in offsets is limited: abs(offset0 - offset1) < 2^31.
-  int cmp(const VPointer& q) const {
-    if (valid() && q.valid() &&
-        (_adr == q._adr || (_base == _adr && q._base == q._adr)) &&
-        _scale == q._scale   && invar_equals(q)) {
-      jlong difference = abs(java_subtract((jlong)_offset, (jlong)q._offset));
-      jlong max_diff = (jlong)1 << 31;
-      if (difference >= max_diff) {
-        return NotComparable;
+  bool has_same_invar_and_iv_scale_as(const VPointer& other) const {
+    // If we have the same invar_summands, and the same iv summand with the same iv_scale,
+    // then all summands except the base must be the same.
+    return mem_pointer().has_same_non_base_summands_as(other.mem_pointer());
+  }
+
+  bool is_adjacent_to_and_before(const VPointer& other) const {
+    if (!is_valid() || !other.is_valid()) {
+#ifndef PRODUCT
+      if (_vloop.mptrace().is_trace_overlap()) {
+        tty->print_cr("VPointer::is_adjacent_to_and_before: invalid VPointer, adjacency unknown.");
       }
-      bool overlap = q._offset <   _offset +   memory_size() &&
-                       _offset < q._offset + q.memory_size();
-      return overlap ? Equal : (_offset < q._offset ? Less : Greater);
-    } else {
-      return NotComparable;
+#endif
+      return false;
     }
+    return mem_pointer().is_adjacent_to_and_before(other.mem_pointer());
   }
 
-  bool overlap_possible_with_any_in(const GrowableArray<Node*>& nodes) const {
-    for (int i = 0; i < nodes.length(); i++) {
-      MemNode* mem = nodes.at(i)->as_Mem();
-      VPointer p_mem(mem, _vloop);
-      // Only if we know that we have Less or Greater can we
-      // be sure that there can never be an overlap between
-      // the two memory regions.
-      if (!not_equal(p_mem)) {
-        return true;
```

Comment:
Note: refactored away. We used to pass in all nodes in a pack/vector, find all their `VPointer`. Now the vector has its own size-adjusted `VPointer`, and we can simply do a `never_overlaps_with` query directly on that.